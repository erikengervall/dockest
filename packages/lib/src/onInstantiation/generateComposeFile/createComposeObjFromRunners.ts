import { ComposeService, DependsOn, ComposeFile } from '../../runners/@types'
import { DockestConfig } from '../../index'

const { keys } = Object

const getDepComposeServices = (dependsOn: DependsOn) =>
  dependsOn.reduce(
    (composeServices: { [key: string]: ComposeService }, { runnerConfig: { service }, getComposeService }) => ({
      ...composeServices,
      [service]: getComposeService(),
    }),
    {},
  )

export default (config: DockestConfig, dockerComposeFileVersion: string) => {
  const composeObj: ComposeFile = {
    version: dockerComposeFileVersion,
    services: {},
  }

  config.runners.forEach(runner => {
    const {
      runnerConfig: { service, dependsOn },
      getComposeService,
    } = runner

    const composeService = getComposeService()
    const depComposeServices = getDepComposeServices(dependsOn)

    composeObj.services = {
      ...composeObj.services,
      [service]: { ...composeService },
      ...depComposeServices,
    }
  })

  const servicesWithNetworks = keys(composeObj.services).filter(
    service => composeObj.services[service].networks !== undefined,
  )
  if (servicesWithNetworks.length > 0) {
    composeObj.networks = servicesWithNetworks.reduce(
      (networks, service) => ({
        ...networks,
        ...composeObj.services[service].networks,
      }),
      {},
    )
  }

  const servicesWithVolumes = keys(composeObj.services).filter(
    service => composeObj.services[service].volumes !== undefined,
  )
  if (servicesWithVolumes.length > 0) {
    composeObj.volumes = servicesWithVolumes.reduce(
      (volumes, service) => ({
        ...volumes,
        ...composeObj.services[service].volumes,
      }),
      {},
    )
  }

  return composeObj
}
