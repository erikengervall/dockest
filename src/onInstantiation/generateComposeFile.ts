import { default as fsLib } from 'fs'
import { default as yamlLib } from 'js-yaml'
import { ComposeService, DependsOn } from '../runners/@types'
import { DockestConfig } from '../index'
import DockestError from '../errors/DockestError'

const { keys } = Object

export interface ComposeFile {
  version: string
  services: {
    [key: string]: ComposeService
  }
  networks?: {
    [key: string]: null
  }
  volumes?: {
    [key: string]: null
  }
}

const getDepComposeServices = (dependsOn: DependsOn) =>
  dependsOn.reduce(
    (composeServices: { [key: string]: ComposeService }, { runnerConfig: { service }, getComposeService }) => ({
      ...composeServices,
      [service]: getComposeService(),
    }),
    {},
  )

export default (config: DockestConfig, yaml = yamlLib, fs = fsLib) => {
  const composeFile: ComposeFile = {
    version: '3',
    services: {},
  }

  config.runners.forEach(runner => {
    const {
      runnerConfig: { service, dependsOn },
      getComposeService,
    } = runner

    const composeService = getComposeService()
    const depComposeServices = getDepComposeServices(dependsOn)

    composeFile.services = {
      ...composeFile.services,
      [service]: composeService,
      ...depComposeServices,
    }
  })

  const servicesWithNetworks = keys(composeFile.services).filter(
    service => composeFile.services[service].networks !== undefined,
  )
  if (servicesWithNetworks.length > 0) {
    composeFile.networks = servicesWithNetworks.reduce(
      (networks, service) => ({
        ...networks,
        ...composeFile.services[service].networks,
      }),
      {},
    )
  }

  const servicesWithVolumes = keys(composeFile.services).filter(
    service => composeFile.services[service].volumes !== undefined,
  )
  if (servicesWithVolumes.length > 0) {
    composeFile.volumes = servicesWithVolumes.reduce(
      (volumes, service) => ({
        ...volumes,
        ...composeFile.services[service].volumes,
      }),
      {},
    )
  }
  const yml = yaml.safeDump(composeFile)

  try {
    fs.writeFileSync(`${process.cwd()}/docker-compose-dockest.yml`, yml)
  } catch (error) {
    throw new DockestError(`Something went wrong when generating the docker-compose file: ${error.message}`)
  }
}
