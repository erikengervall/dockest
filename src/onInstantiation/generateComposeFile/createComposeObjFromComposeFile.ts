import { default as fsLib } from 'fs'
import { mergeDeepRight } from 'ramda'
import { default as yamlLib } from 'js-yaml'
import { DockestConfig } from '../../index'
import { ComposeFile } from '../../runners/@types'

export default (config: DockestConfig, fs = fsLib, yaml = yamlLib, nodeProcess = process): ComposeFile => {
  const getComposeObjFromComposeFile = (fileName: string) => {
    const composeYml = fs.readFileSync(`${nodeProcess.cwd()}/${fileName}`, 'utf8')
    const composeObj: ComposeFile = yaml.safeLoad(composeYml)

    return composeObj
  }

  let composeObj = {
    version: '3',
    services: {},
  }

  const { composeFile } = config.opts
  if (composeFile.length > 0) {
    const composeObjsFromFile = Array.isArray(composeFile)
      ? composeFile.map(fileName => getComposeObjFromComposeFile(fileName))
      : [getComposeObjFromComposeFile(composeFile)]

    composeObjsFromFile.forEach(composeObjFromFile => {
      composeObj = mergeDeepRight(composeObj, composeObjFromFile)
    })
  }

  return composeObj
}
