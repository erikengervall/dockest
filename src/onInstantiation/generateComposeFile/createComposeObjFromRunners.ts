import execa from 'execa'
import { ComposeService, DependsOn, ComposeFile } from '../../runners/@types'
import { DockestConfig } from '../../index'

const { keys } = Object

const getDepComposeServices = (dependsOn: DependsOn) =>
  dependsOn.reduce(
    (composeServices: { [key: string]: ComposeService }, { runnerConfig: { service }, getComposeService }) => ({
      ...composeServices,
      [service]: getComposeService(),
    }),
    {},
  )

export default (config: DockestConfig, dockerComposeFileVersion: string) => {
  const composeObj: ComposeFile = {
    version: dockerComposeFileVersion,
    services: {},
  }

  const extra_hosts: string[] = []

  if (!config.$.isInsideDockerContainer) {
    if (process.platform === 'linux') {
      const command = `ip -4 addr show docker0 | grep -oP '(?<=inet\s)\d+(\.\d+){3}'`
      const result = execa.sync(command, { reject: false })
      extra_hosts.push(`host.docker.internal:${result.stdout}`)
    }
  }

  config.runners.forEach(runner => {
    const {
      runnerConfig: { service, dependsOn },
      getComposeService,
    } = runner

    const composeService = getComposeService()
    const depComposeServices = getDepComposeServices(dependsOn)

    composeObj.services = {
      ...composeObj.services,
      [service]: { ...composeService, extra_hosts },
      ...depComposeServices,
    }
  })

  const servicesWithNetworks = keys(composeObj.services).filter(
    service => composeObj.services[service].networks !== undefined,
  )
  if (servicesWithNetworks.length > 0) {
    composeObj.networks = servicesWithNetworks.reduce(
      (networks, service) => ({
        ...networks,
        ...composeObj.services[service].networks,
      }),
      {},
    )
  }

  const servicesWithVolumes = keys(composeObj.services).filter(
    service => composeObj.services[service].volumes !== undefined,
  )
  if (servicesWithVolumes.length > 0) {
    composeObj.volumes = servicesWithVolumes.reduce(
      (volumes, service) => ({
        ...volumes,
        ...composeObj.services[service].volumes,
      }),
      {},
    )
  }

  return composeObj
}
