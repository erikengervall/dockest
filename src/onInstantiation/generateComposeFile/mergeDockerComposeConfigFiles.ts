import execa from 'execa'
import path from 'path'
import getComposeObjFromComposeYmlString from './getComposeObjFromComposeYmlString'

export default (cwd: string, composeFiles: string[]) => {
  const result = execa.sync(
    'docker-compose',
    composeFiles
      .slice()
      .reverse()
      .reduce(
        (result, composeFilePath) => {
          result.unshift('-f', path.join(cwd, composeFilePath))
          return result
        },
        ['config'],
      ),
    {
      reject: false,
    },
  )

  if (result.exitCode !== 0) {
    console.error(
      `ðŸš¨ Invalid docker-compose config: \n ${result.stderr}\n\n You can declare the given option via your runner definition or preferably the compose file. \n`,
    )
    throw new TypeError('Invalid docker-compose config.')
  }

  return getComposeObjFromComposeYmlString(result.stdout)
}
