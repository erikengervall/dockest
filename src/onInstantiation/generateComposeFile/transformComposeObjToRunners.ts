import { RedisRunner, PostgresRunner, GeneralPurposeRunner, ZooKeeperRunner, KafkaRunner } from '../../runners'
import { DockestConfig } from '../../index'
import { ComposeFile } from '../../runners/@types'
import ConfigurationError from '../../errors/ConfigurationError'

const { keys } = Object

const REDIS_REG_EXP = RegExp('redis')
const POSTGRES_REG_EXP = RegExp('postgres')
const ZOOKEEPER_REG_EXP = RegExp('zookeeper')
const KAFKA_REG_EXP = RegExp('kafka')

export default (config: DockestConfig, composeObj: ComposeFile) =>
  keys(composeObj.services).map(service => {
    const attachedRunnerExist = config.runners.find(runner => runner.runnerConfig.service === service)
    if (attachedRunnerExist) {
      return attachedRunnerExist
    }

    const composeService = composeObj.services[service]

    /**
     * Redis
     */
    if (REDIS_REG_EXP.test(composeService.image || '') || REDIS_REG_EXP.test(service)) {
      return new RedisRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
      })
    }

    /**
     * Postgres
     */
    if (POSTGRES_REG_EXP.test(composeService.image || '') || POSTGRES_REG_EXP.test(service)) {
      if (!composeService.environment) {
        throw new ConfigurationError(`${service}: Invalid environment`)
      }

      return new PostgresRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
        database: `${composeService.environment[PostgresRunner.ENVIRONMENT_DATABASE]}`,
        password: `${composeService.environment[PostgresRunner.ENVIRONMENT_PASSWORD]}`,
        username: `${composeService.environment[PostgresRunner.ENVIRONMENT_USERNAME]}`,
      })
    }

    /**
     * ZooKeeper
     */
    if (ZOOKEEPER_REG_EXP.test(composeService.image || '') || ZOOKEEPER_REG_EXP.test(service)) {
      if (!composeService.environment) {
        throw new ConfigurationError(`${service}: Invalid environment`)
      }

      return new ZooKeeperRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
      })
    }

    /**
     * Kafka
     */
    if (KAFKA_REG_EXP.test(composeService.image || '') || KAFKA_REG_EXP.test(service)) {
      if (!composeService.environment) {
        throw new ConfigurationError(`${service}: Invalid environment`)
      }

      return new KafkaRunner({
        service,
        image: composeService.image,
        ports: composeService.ports,
        props: composeService.environment,
      })
    }

    return new GeneralPurposeRunner({
      service,
      image: composeService.image,
    })
  })
