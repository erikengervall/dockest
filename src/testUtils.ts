import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
