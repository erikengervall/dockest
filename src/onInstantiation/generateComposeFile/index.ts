import { default as fsLib } from 'fs'
import { default as yamlLib } from 'js-yaml'
import { mergeDeepRight } from 'ramda'
import createComposeObjFromComposeFile from './createComposeObjFromComposeFile'
import createComposeObjFromRunners from './createComposeObjFromRunners'
import transformComposeObjToRunners from './transformComposeObjToRunners'
import { DockestConfig } from '../../index'
import { GENERATED_COMPOSE_FILE_PATH } from '../../constants'

export default (config: DockestConfig, yaml = yamlLib, fs = fsLib) => {
  const composeObjFromComposeFile = createComposeObjFromComposeFile(config)
  const composeObjFromRunners = createComposeObjFromRunners(config)

  const composeObj = mergeDeepRight(composeObjFromComposeFile, composeObjFromRunners)
  config.runners = transformComposeObjToRunners(config, composeObj)

  fs.writeFileSync(GENERATED_COMPOSE_FILE_PATH, yaml.safeDump(composeObj))
}
