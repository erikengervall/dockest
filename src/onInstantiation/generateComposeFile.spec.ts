/* eslint-disable @typescript-eslint/ban-ts-ignore */

import yaml from 'js-yaml'
import generateComposeFile from './generateComposeFile'
import testUtils from '../testUtils'

const {
  createDockestConfig,
  initializedRunners,
  runners: { GeneralPurposeRunner },
} = testUtils({})

describe('generateComposeFile', () => {
  const yamlMock: any = {
    safeDump: jest.fn(json => yaml.safeDump(json)),
  }
  const fsMock: any = {
    writeFileSync: jest.fn(),
  }

  beforeEach(() => {
    yamlMock.safeDump.mockClear()
    fsMock.writeFileSync.mockClear()
  })

  it('should handle the standard initializedRunners', () => {
    const dockestConfig = createDockestConfig({
      runners: Object.values(initializedRunners),
    })

    generateComposeFile(dockestConfig, yamlMock, fsMock)

    expect(yamlMock.safeDump).toMatchSnapshot()
    expect(fsMock.writeFileSync).toHaveBeenCalledTimes(1)
  })

  it('should handle runners with networks prop', () => {
    const dockestConfig = createDockestConfig({
      runners: [
        new GeneralPurposeRunner({
          service: 'networks1',
          image: 'some/thing:101',
          networks: ['bueno'],
          ports: { networks1: 'networks1' },
        }),
        new GeneralPurposeRunner({
          service: 'networks2',
          image: 'some/thing:101',
          networks: ['bueno'],
          ports: { networks2: 'networks2' },
        }),
      ],
    })

    generateComposeFile(dockestConfig, yamlMock, fsMock)

    expect(yamlMock.safeDump).toMatchSnapshot()
    expect(fsMock.writeFileSync).toHaveBeenCalledTimes(1)
  })

  it('should handle runners with build prop', () => {
    const dockestConfig = createDockestConfig({
      runners: [
        new GeneralPurposeRunner({
          service: 'build1',
          build: './some/path',
          ports: { build1: 'build1' },
        }),
        new GeneralPurposeRunner({
          service: 'build2',
          build: './some/path',
          ports: { build1: 'build2' },
        }),
      ],
    })

    generateComposeFile(dockestConfig, yamlMock, fsMock)

    expect(yamlMock.safeDump).toMatchSnapshot()
    expect(fsMock.writeFileSync).toHaveBeenCalledTimes(1)
  })
})
