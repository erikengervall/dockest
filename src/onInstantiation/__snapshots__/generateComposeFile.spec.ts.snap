// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateComposeFile should handle runners with build prop 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "services": Object {
          "build1": Object {
            "build": "./some/path",
            "ports": Array [
              "build1:build1",
            ],
          },
          "build2": Object {
            "build": "./some/path",
            "ports": Array [
              "build1:build2",
            ],
          },
        },
        "version": "3",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "version: '3'
services:
  build1:
    build: ./some/path
    ports:
      - 'build1:build1'
  build2:
    build: ./some/path
    ports:
      - 'build1:build2'
",
    },
  ],
}
`;

exports[`generateComposeFile should handle runners with networks prop 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "networks": Object {
          "bueno": null,
        },
        "services": Object {
          "networks1": Object {
            "image": "some/thing:101",
            "networks": Object {
              "bueno": null,
            },
            "ports": Array [
              "networks1:networks1",
            ],
          },
          "networks2": Object {
            "image": "some/thing:101",
            "networks": Object {
              "bueno": null,
            },
            "ports": Array [
              "networks2:networks2",
            ],
          },
        },
        "version": "3",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "version: '3'
services:
  networks1:
    image: 'some/thing:101'
    networks:
      bueno: null
    ports:
      - 'networks1:networks1'
  networks2:
    image: 'some/thing:101'
    networks:
      bueno: null
    ports:
      - 'networks2:networks2'
networks:
  bueno: null
",
    },
  ],
}
`;

exports[`generateComposeFile should handle the standard initializedRunners 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "services": Object {
          "general": Object {
            "image": "some/image:123",
            "ports": Array [],
          },
          "kafka": Object {
            "depends_on": Array [
              "zookeepeer",
            ],
            "environment": Object {
              "KAFKA_ADVERTISED_LISTENERS": "PLAINTEXT://kafka:29092,PLAINTEXT_HOST://localhost:9092",
              "KAFKA_AUTO_CREATE_TOPICS_ENABLE": "true",
              "KAFKA_BROKER_ID": 1,
              "KAFKA_LISTENER_SECURITY_PROTOCOL_MAP": "PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT",
              "KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR": 1,
              "KAFKA_ZOOKEEPER_CONNECT": "zookeepeer:2181",
            },
            "image": "some/image:123",
            "ports": Array [
              "9092:9092",
            ],
          },
          "postgres": Object {
            "environment": Object {
              "POSTGRES_DB": "_",
              "POSTGRES_PASSWORD": "_",
              "POSTGRES_USER": "_",
            },
            "image": "some/image:123",
            "ports": Array [
              "5432:5432",
            ],
          },
          "redis": Object {
            "image": "some/image:123",
            "ports": Array [
              "6379:6379",
            ],
          },
          "zookeepeer": Object {
            "environment": Object {
              "ZOOKEEPER_CLIENT_PORT": "2181",
            },
            "image": "some/image:123",
            "ports": Array [
              "2181:2181",
            ],
          },
        },
        "version": "3",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "version: '3'
services:
  kafka:
    environment:
      KAFKA_ZOOKEEPER_CONNECT: 'zookeepeer:2181'
      KAFKA_ADVERTISED_LISTENERS: 'PLAINTEXT://kafka:29092,PLAINTEXT_HOST://localhost:9092'
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: 'PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT'
      KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'true'
      KAFKA_BROKER_ID: 1
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
    depends_on:
      - zookeepeer
    image: 'some/image:123'
    ports:
      - '9092:9092'
  zookeepeer:
    environment:
      ZOOKEEPER_CLIENT_PORT: '2181'
    image: 'some/image:123'
    ports:
      - '2181:2181'
  postgres:
    environment:
      POSTGRES_DB: _
      POSTGRES_PASSWORD: _
      POSTGRES_USER: _
    image: 'some/image:123'
    ports:
      - '5432:5432'
  redis:
    image: 'some/image:123'
    ports:
      - '6379:6379'
  general:
    image: 'some/image:123'
    ports: []
",
    },
  ],
}
`;
